mat64/gsvd.go

```diff
@@ -0,0 +1,296 @@
+	}
+	if c == nil {
+		c = make([]float64, min(gsvd.m, gsvd.n)-gsvd.k)
+	}
+	if len(c) != min(gsvd.m, gsvd.n)-gsvd.k {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(c, gsvd.c[gsvd.k:min(gsvd.m, gsvd.n)])
+	return c
+}
+
+// Values2 returns the singular values of the factorized matrix.
+// If the input slice is non-nil, the values will be stored in-place into the slice.
+// In this case, the slice must have length min(m,n)-k, and Values2 will panic with
+// matrix.ErrSliceLengthMismatch otherwise. If the input slice is nil,
+// a new slice of the appropriate length will be allocated and returned.
+//
+// Values2 will panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) Values2(s []float64) []float64 {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	if s == nil {
+		s = make([]float64, min(gsvd.m, gsvd.n)-gsvd.k)
+	}
+	if len(s) != min(gsvd.m, gsvd.n)-gsvd.k {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(s, gsvd.s[gsvd.k:min(gsvd.m, gsvd.n)])
+	return s
+}
+
+// ZeroRFromGSVD extracts the matrix [ 0 R ] from the singular value decomposition, storing
+// the result in-place into the receiver. [ 0 R ] is size (k+l)×n.
+func (m *Dense) ZeroRFromGSVD(gsvd *GSVD) {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	m.reuseAsZeroed(gsvd.k+gsvd.l, gsvd.n)
+	a := Dense{
+		mat:     gsvd.a,
+		capRows: gsvd.m,
+		capCols: gsvd.n,
+	}
+	m.Slice(0, min(gsvd.k+gsvd.l, gsvd.m), gsvd.n-gsvd.k-gsvd.l, gsvd.n).(*Dense).
+		Copy(a.Slice(0, min(gsvd.k+gsvd.l, gsvd.m), gsvd.n-gsvd.k-gsvd.l, gsvd.n))
+	if gsvd.m < gsvd.k+gsvd.l {
+		b := Dense{
+			mat:     gsvd.b,
+			capRows: gsvd.p,
+			capCols: gsvd.n,
+		}
+		m.Slice(gsvd.m, gsvd.k+gsvd.l, gsvd.n+gsvd.m-gsvd.k-gsvd.l, gsvd.n).(*Dense).
+			Copy(b.Slice(gsvd.m-gsvd.k, gsvd.l, gsvd.n+gsvd.m-gsvd.k-gsvd.l, gsvd.n))
+	}
+}
+
+// Sigma1FromGSVD extracts the matrix Σ₁ from the singular value decomposition, storing
+// the result in-place into the receiver. Σ₁ is size m×(k+l).
+func (m *Dense) Sigma1FromGSVD(gsvd *GSVD) {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	m.reuseAsZeroed(gsvd.m, gsvd.k+gsvd.l)
+	for i := 0; i < gsvd.k; i++ {
+		m.set(i, i, 1)
+	}
+	for i := gsvd.k; i < min(gsvd.m, gsvd.k+gsvd.l); i++ {
+		m.set(i, i, gsvd.c[i])
+	}
+}
+
+// Sigma2FromGSVD extracts the matrix Σ₂ from the singular value decomposition, storing
+// the result in-place into the receiver. Σ₂ is size p×(k+l).
+func (m *Dense) Sigma2FromGSVD(gsvd *GSVD) {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	m.reuseAsZeroed(gsvd.p, gsvd.k+gsvd.l)
+	for i := 0; i < min(gsvd.l, gsvd.m-gsvd.k); i++ {
+		m.set(i, i+gsvd.k, gsvd.s[gsvd.k+i])
+	}
+	for i := gsvd.m - gsvd.k; i < gsvd.l; i++ {
+		m.set(i, i+gsvd.k, 1)
+	}
+}
+
+// UFromGSVD extracts the matrix U from the singular value decomposition, storing
+// the result in-place into the receiver. U is size m×m.
+func (m *Dense) UFromGSVD(gsvd *GSVD) {
```

Comment:
This PR predates #436. When #436 is merged, this PR will be updated to follow the pattern set there,